temporal-testing/src/main/java/io/temporal/internal/testservice/TestWorkflowMutableStateImpl.java

```diff
@@ -2115,6 +2126,209 @@ public void completeQuery(QueryId queryId, RespondQueryTaskCompletedRequest comp
+        .setWorkflowExecutionInfo(executionInfo)
+        .addAllPendingActivities(pendingActivities)
+        .addAllPendingChildren(pendingChildren)
+        .build();
+  }
+
+  private PendingChildExecutionInfo constructPendingChildExecutionInfo(
+      StateMachine<ChildWorkflowData> sm) {
+    ChildWorkflowData data = sm.getData();
+    return PendingChildExecutionInfo.newBuilder()
+        .setWorkflowId(data.execution.getWorkflowId())
+        .setRunId(data.execution.getRunId())
+        .setWorkflowTypeName(data.initiatedEvent.getWorkflowType().getName())
+        .setInitiatedId(data.initiatedEventId)
+        .setParentClosePolicy(data.initiatedEvent.getParentClosePolicy())
+        .build();
+  }
+
+  private PendingActivityInfo constructPendingActivityInfo(StateMachine<ActivityTaskData> sm) {
+    /*
+     * Working on this code? First, go read StateMachines.scheduleActivityTask to learn which
+     * information it puts in pendingActivity.scheduledEvent and which information it puts
+     * in pendingActivity.activityTask. The ActivityTaskData itself also has some of what we need.
+     */
+    ActivityTaskData pendingActivity = sm.getData();
+    PollActivityTaskQueueResponse.Builder task = pendingActivity.activityTask.getTask();
+    ActivityTaskScheduledEventAttributes scheduledEvent = pendingActivity.scheduledEvent;
+
+    State state = sm.getState();
+    PendingActivityInfo.Builder builder =
+        PendingActivityInfo.newBuilder()
+            .setActivityId(scheduledEvent.getActivityId())
+            .setActivityType(scheduledEvent.getActivityType())
+            .setState(computeActivityState(state, pendingActivity));
+
+    // In golang, we set one but never both of these fields, depending on the activity state
+    if (builder.getState() == PendingActivityState.PENDING_ACTIVITY_STATE_SCHEDULED) {
+      builder.setScheduledTime(task.getScheduledTime());
+    } else {
+      builder.setLastStartedTime(task.getStartedTime());
+      builder.setLastHeartbeatTime(task.getStartedTime());
+    }
+
+    if (pendingActivity.lastHeartbeatTime > 0) {
+      // This may overwrite the heartbeat time we just set - that's fine
+      builder.setLastHeartbeatTime(timestampFromMillis(pendingActivity.lastHeartbeatTime));
+
+      if (pendingActivity.heartbeatDetails != null) {
+        builder.setHeartbeatDetails(pendingActivity.heartbeatDetails);
+      }
+    }
+
+    if (scheduledEvent.hasRetryPolicy()) {
+      builder.setAttempt(pendingActivity.retryState.getAttempt());
+      builder.setExpirationTime(pendingActivity.retryState.getExpirationTime());
+      builder.setMaximumAttempts(scheduledEvent.getRetryPolicy().getMaximumAttempts());
+      pendingActivity.retryState.getPreviousRunFailure().ifPresent(builder::setLastFailure);
+      // We don't track this in the test environment right now, but we could.
+      builder.setLastWorkerIdentity("test-environment-worker-identity");
+    } else {
+      builder.setAttempt(1);
```

Comment:
If you replace `builder.setAttempt(pendingActivity.retryState.getAttempt());` with `builder.setAttempt(pendingActivity.getAttempt());` and move it from the if statement - you can kill this else completely, because `retryState != null ? retryState.getAttempt() : 1` logic is already located inside `pendingActivity.getAttempt()`

Reply:
I think I can clean this up even further - in `StateMachines.scheduleActivityTask`, `.retryState` is always set and the attempt number starts at 1 and gets mutated from there. 

I'm 99% sure retryState is never null and retryPolicy is never null (but may have been in very old schemas, because the golang code makes allowances for those cases). Let's confirm/deny over slack.